packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -479,6 +478,103 @@ void Add(
   }
 }
 
+namespace ColMapFunctors
+{
+  template<typename ByteView, typename GView>
+  struct UnionEntries
+  {
+    UnionEntries(ByteView& entryUnion_, const GView& gids_) : entryUnion(entryUnion_), gids(gids_) {}
+    KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+    {
+      entryUnion(gids(i)) = 1;
+    }
+    ByteView& entryUnion;
+    const GView& gids;
+  };
+
+  template<typename ByteView, typename LView>
+  struct EntriesPrefixSum
+  {
+    EntriesPrefixSum(const ByteView& entryUnion_, LView& gtol_) : entryUnion(entryUnion_), gtol(gtol_) {}
+    KOKKOS_INLINE_FUNCTION void operator()(const size_t i, size_t& update, const bool finalPass) const
+    {
+      // exclusive scan
+      if(finalPass)
+      {
+        gtol(i) = update;
+      }
+      if(entryUnion(i))
+      {
+        update++;
+      }
+      if(finalPass)
+      {
+        if(i == entryUnion.dimension_0() - 1)
+        {
+          gtol(i + 1) = update;
+        }
+      }
+    }
+    const ByteView& entryUnion;
+    LView& gtol;
+  };
+
+  template<typename LView, typename GView>
+  struct ConvertGlobalToLocal
+  {
+    ConvertGlobalToLocal(const LView& gtol_, const GView& gids_, LView& lids_) : gtol(gtol_), gids(gids_), lids(lids_) {}
+    KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+    {
+      lids(i) = gtol(gids(i));
+    }
+    const LView& gtol;
+    const GView& gids;
+    LView& lids;
+  };
+}
+
+//Build the minimal (sorted) column map for the given set of global columns
+//Then convert gids and store them in lids (gids is not modified)
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+Teuchos::RCP<Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>> AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+getOptimizedColMap(GlobalOrdinal ncols,
+                   const typename AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GlobalColIndsArray& gids,
+                   typename AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::ColIndsArray& lids,
+                   const Teuchos::RCP<const Teuchos::Comm<int>>& comm)
+{
+  using namespace ColMapFunctors;
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  typedef Kokkos::View<char*, device_type> ByteView;
+  typedef typename AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GlobalColIndsArray GView;
+  typedef typename AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::ColIndsArray LView;
+  //Functors (explained in the procedural code below)
+  auto nentries = gids.dimension_0();
+  //each entry of entryUnion is 0 unless there is a local entry in that column (then it is 1)
```

Comment:
@mhoemmen I'm thinking that the allocation time and extra memory usage here is acceptable, because rowptrs for each matrix is the same size (at least 3 of those exist during add()), and values/colinds are at least that big (even if only 1-2 entries per row).